File: python/111.二叉树的最小深度.py
#
# @lc app=leetcode.cn id=111 lang=python3
#
# [111] 二叉树的最小深度
#
# 解法1(T81% S89%)：与最大深度不同，不需要完整遍历，只需要找到第一个叶节点也就找到了最小深度，比较适合用层次遍历，而且在入队的同时将所在level和val打包成元组一同压，当遇到弹出队头的某个元素是叶就直接返回其level即可

# @lc code=start
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def minDepth(self, root: TreeNode) -> int:
        if not root: return 0

        queue = [(root,1)]
        while queue:
            node, level = queue.pop(0)
            if not node.left and not node.right: return level
            if node.left: queue.append((node.left, level+1))
            if node.right: queue.append((node.right, level+1))
        
# @lc code=end

File: javascript/111.二叉树的最小深度.js
/*
 * @lc app=leetcode.cn id=111 lang=javascript
 *
 * [111] 二叉树的最小深度
 * 
 * 最大深度可以直接递归，最小深度只能用层序来找第一个叶子节点
 * 因为只有叶子节点才能说深度是多少，如果还是递归的话，看到一个空子树也会直接返回最小的
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number}
 */
var minDepth = function(root) {
    if (!root) { return 0; }
    var queue = [[root, 1]];
    while (queue) {
        var data = queue.shift();
        var node = data[0], level = data[1];
        if (!node.left && !node.right) { return level; }
        if (node.left) {
            queue.push([node.left, level+1]);
        }
        if (node.right) {
            queue.push([node.right, level+1]);
        }
    }
};
// @lc code=end

